messaging/src/main/java/io/lantern/messaging/Worker.kt

```diff
@@ -0,0 +1,68 @@
+package io.lantern.messaging
+
+import mu.KotlinLogging
+import java.io.Closeable
+import java.util.concurrent.Callable
+import java.util.concurrent.ExecutionException
+import java.util.concurrent.Executors
+import java.util.concurrent.LinkedBlockingQueue
+import java.util.concurrent.TimeUnit
+
+internal abstract class Worker(
+    protected val messaging: Messaging,
+    name: String,
+    retryDelayMillis: Long? = null
+) : Closeable {
+    protected val logger = KotlinLogging.logger("${messaging.logger.name}-$name")
+
+    internal val executor = Executors.newSingleThreadScheduledExecutor {
+        Thread(it, "${messaging.name}-$name-executor")
+    }
+
+    private val retries = LinkedBlockingQueue<() -> Unit>()
+
+    init {
+        if (retryDelayMillis != null) {
+            logger.debug("will automatically retry every ${retryDelayMillis}ms")
+            executor.scheduleAtFixedRate(
+                {
+                    while (true) {
+                        retries.poll()?.let { submit(it) } ?: return@scheduleAtFixedRate
+                    }
+                },
+                retryDelayMillis, retryDelayMillis, TimeUnit.MILLISECONDS
+            )
+        }
+    }
+
+    internal fun submit(cmd: () -> Unit) {
+        try {
+            executor.submit {
+                try {
+                    cmd()
+                } catch (t: Throwable) {
+                    logger.error(t.message, t)
+                    retryFailed(cmd)
+                }
+            }
+        } catch (t: Throwable) {
+            logger.error(t.message)
+        }
+    }
```

Comment:
@oxtoacart Still getting familiar with the codebase, but just curious if there was a particular reason you decided to use the Java ExecutorService API here in this Worker class vs. Kotlin coroutines? https://developer.android.com/kotlin/coroutines/coroutines-adv I haven't played around much with them yet, but my understanding is they're lighter weight than threads. They also introduce a concept called "CoroutineScopes" (not sure if you're familiar with them but they implement structured concurrency and give you an easy way to manage the lifecycle of an asynchronous task; plus, the syntax is much simpler). For instance, you could do something simple like this to implement a repeatable task:

```kotlin
private fun startRepeatingJob(): Job {
  return CoroutineScope(Dispatchers.Default).launch {
    repeat(maxRetries){
      try {
        val cmdAsync = async {
            mayFail()
        }
        useResult(cmdAsync.await() /*,cmdAsync2.await()*/)
       } catch (t: Throwable) {
         logger.error(t.message)
       }
       delay(retryDelayMillis)
    }
  }
```

Anyway, just curious! =)

Reply:
Hehe, I actually started out using coroutines. I ended up running into issues when using some threaded libraries like a networking library that I was using at the time, and I also had issues with the database code.